javascript/ql/src/semmle/javascript/dataflow/Nodes.qll

```diff
@@ -462,3 +462,214 @@ DataFlow::SourceNode moduleMember(string path, string m) {
     result = DataFlow::ssaDefinitionNode(ssa)
   )
 }
+
+/**
+ * A data flow node corresponding to a class definition or a function definition
+ * acting as a class.
+ *
+ * The following patterns are recognized as classes and methods:
+ * ```
+ * class C {
+ *   method()
+ * }
+ *
+ * function F() {}
+ *
+ * F.prototype.method = function() {}
+ *
+ * F.prototype = {
+ *   method: function() {}
+ * }
+ *
+ * extend(F.prototype, {
+ *   method: function() {}
+ * });
+ * ```
+ */
+class ClassNode extends DataFlow::SourceNode {
+  ClassNode::Range impl;
+
+  ClassNode() { this = impl }
+
+  /**
+   * Gets the name of the class, if it has one.
+   */
+  string getName() { result = impl.getName() }
+
+  /**
+   * Gets a description of the class.
+   */
+  string desribe() { result = impl.describe() }
+
+  /**
+   * Gets the constructor function of this class.
+   */
+  FunctionNode getConstructor() { result = impl.getConstructor() }
+
+  /**
+   * Gets an instance method with the given name, if any.
+   */
+  FunctionNode getAnInstanceMethod(string name) { result = impl.getAnInstanceMethod(name) }
+
+  /**
+   * Gets an instance method of this class.
+   *
+   * The constructor is not considered an instance method.
+   */
+  FunctionNode getAnInstanceMethod() { result = impl.getAnInstanceMethod() }
+
+  /**
+   * Gets the static method of this class with the given name.
+   */
+  FunctionNode getAStaticMethod(string name) { result = impl.getAStaticMethod(name) }
+
+  /**
+   * Gets a static method of this class.
+   *
+   * The constructor is not considered a static method.
+   */
+  FunctionNode getAStaticMethod() { result = impl.getAStaticMethod() }
+}
+
+module ClassNode {
+  /**
+   * A dataflow node that should be considered a class node.
+   *
+   * Subclass this to introduce new kinds of class nodes. If you want to refine
+   * the definition of existing class nodes, subclass `DataFlow::ClassNode` instead.
+   */
+  abstract class Range extends DataFlow::SourceNode {
+    /**
+     * Gets the name of the class, if it has one.
+     */
+    abstract string getName();
+
+    /**
+     * Gets a description of the class.
+     */
+    abstract string describe();
+
+    /**
+     * Gets the constructor function of this class.
+     */
+    abstract FunctionNode getConstructor();
+
+    /**
+     * Gets an instance method with the given name, if any.
+     */
+    abstract FunctionNode getAnInstanceMethod(string name);
```

Comment:
I think we should mention that the prototype chains are not traversed for these lookups.
I am not sure if the prototype traversal should be implicit or explicit one day.